### PR TITLE
feat: handling of cairo u512 type

### PR DESCRIPTION
**Pull request by PhilippeR26**
## Motivation and Resolution
The new u512 Cairo type is now implemented.
Solve issue #1013.
 According to https://github.com/starkware-libs/cairo/blob/07484c52791b76abcc18fd86265756904557d0d2/corelib/src/test/integer_test.cairo#L767

The raw format is :
```
{
    limb0: u128, // lower weight
    limb1: u128,
    limb2: u128,
    limb3: u128, // upper weight
};
```

## Usage related changes

User send a BigNumberish, and receive a bigint.

```typescript
await myContract0.my_function(553844998243714947043252949842317834769n);
await myContract1.my_function(cairo.uint512("0xa9d2d1501ad0a2eb5337a9d2d1501ad0a2eb5337a9d2d1501ad0a2eb5337a9d2d1501ad0a2eb5337a9d2d1501ad0a2eb5337")); 
await myContract2.my_function(12345678, '13456789765', '0xe23a40b543f', 1534566734334n);
```

```typescript
const a2: Uint512 = {
  limb0: "0xeb5337d9a885be310x9365205a414fdd",
  limb1: "0x1fd465baff2ba9d2d1501ad0a2eb5337",
  limb2: "0x05f7cd1fd465baff2ba9d2d1501ad0a2",
  limb3: "0x2eb5337d9a885be319366b5205a414fd"
};
```

## Development related changes

A new file is in utils/cairoDataTypes, inspired from the Cairo u256 type.

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Linked the issues which this PR resolves
- [x] Documented the changes in code (API docs will be generated automatically)
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

**Comment by PhilippeR26** [comment]
@haroune-mohammedi 

For u512, Starknet.js is waiting a Bignumberish.

**Review by tabaktoni — CHANGES_REQUESTED** [review]
Look like there is some issue on validate ?

File: www/docs/guides/define_call_message.md [hunk 1]
```diff
@@ -94,6 +94,32 @@ const a2: Uint256 = {
 const a3: Uint256 = { low: a1.low, high: a1.high };
 ```
 
+### u512
+
+Starknet is waiting for 4 u128, the first one has the lowest wight.
```
Comment (ivpavici) [inline www/docs/guides/define_call_message.md:99]: wight?

**Comment by PhilippeR26** [comment]
> Look like there is some issue on validate ?

Problem from pr #1007

Solved.

**Review by tabaktoni — APPROVED** [review]
LGTM!

File: src/utils/cairoDataTypes/uint512.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+  public limb2: bigint;
+
+  public limb3: bigint;
+
+  static abiSelector = 'core::integer::u512';
+
+  /**
+   * Default constructor (Lib usage)
+   * @param bigNumberish BigNumberish value representing uin256
```
Comment (ivpavici) [inline src/utils/cairoDataTypes/uint512.ts:28]:
```suggestion
   * @param bigNumberish BigNumberish value representing uint256
```

Comment (PhilippeR26) [reply src/utils/cairoDataTypes/uint512.ts:28]: In fact : `u512`
Corrected

**Review by ivpavici — APPROVED** [review]

**Comment by haroune-mohammedi** [comment]
> @haroune-mohammedi
> 
> For u512, Starknet.js is waiting a Bignumberish.

Will be addressed here https://github.com/keep-starknet-strange/abi-wan-kanabi/issues/68

**Pull request merged**